From daed68cd98ab4a638e4290b4e530e24b51e9830e Mon Sep 17 00:00:00 2001
From: Philippe Normand <pnormand@igalia.com>
Date: Thu, 1 Jul 2010 10:56:19 +0200
Subject: [PATCH] [dshowvideosink] update renderer aspect ratio after force-aspect-ratio
 property update.

A new virtual method has been added to the RendererSupport class. It
is called during the initial renderer configuration and each time the
force-aspect-ratio property of the sink is updated.

Fixes bug 623272.
---
 sys/dshowvideosink/dshowvideosink.cpp |   14 ++++++++++++--
 sys/dshowvideosink/dshowvideosink.h   |    1 +
 2 files changed, 13 insertions(+), 2 deletions(-)

diff --git a/sys/dshowvideosink/dshowvideosink.cpp b/sys/dshowvideosink/dshowvideosink.cpp
index ab9aecd..7a14664 100644
--- a/sys/dshowvideosink/dshowvideosink.cpp
+++ b/sys/dshowvideosink/dshowvideosink.cpp
@@ -256,6 +256,8 @@ gst_dshowvideosink_set_property (GObject * object, guint prop_id,
       break;
     case PROP_KEEP_ASPECT_RATIO:
       sink->keep_aspect_ratio = g_value_get_boolean (value);
+      if (sink->renderersupport)
+        sink->renderersupport->SetAspectRatioMode();
       break;
     case PROP_FULL_SCREEN:
       sink->full_screen = g_value_get_boolean (value);
@@ -979,13 +981,17 @@ public:
       return FALSE;
     }
 
+    SetAspectRatioMode();
+    return TRUE;
+  }
+
+  void SetAspectRatioMode() {
     if (sink->keep_aspect_ratio) {
       control->SetAspectRatioMode(VMR9ARMode_LetterBox);
     }
     else {
       control->SetAspectRatioMode(VMR9ARMode_None);
     }
-    return TRUE;
   }
 
   gboolean SetRendererWindow(HWND window) {
@@ -1102,13 +1108,17 @@ public:
       return FALSE;
     }
 
+    SetAspectRatioMode();
+    return TRUE;
+  }
+
+  void SetAspectRatioMode() {
     if (sink->keep_aspect_ratio) {
       control->SetAspectRatioMode(VMR_ARMODE_LETTER_BOX);
     }
     else {
       control->SetAspectRatioMode(VMR_ARMODE_NONE);
     }
-    return TRUE;
   }
 
   gboolean SetRendererWindow(HWND window) {
diff --git a/sys/dshowvideosink/dshowvideosink.h b/sys/dshowvideosink/dshowvideosink.h
index 9c9d073..8b60087 100644
--- a/sys/dshowvideosink/dshowvideosink.h
+++ b/sys/dshowvideosink/dshowvideosink.h
@@ -54,6 +54,7 @@ public:
   virtual void MoveWindow() = 0;
   virtual void DestroyWindow() = 0;
   virtual void DisplayModeChanged() = 0;
+  virtual void SetAspectRatioMode() = 0;
 };
 
 struct _GstDshowVideoSink
-- 
1.7.1